### PR TITLE
Adding properties for ASM2d system

### PR DESCRIPTION
**Pull request by andrewlee94**
## Summary/Motivation:
This PR adds the necessary property packages for modeling systems using ASM2d kinetics.

A few notes:

1. ASM2d tracks TSS and this is consumed in a number of reactions. However, the reaction kinetics do not depend on TSS and thus is is possible to define a system with insufficient TSS which would result in a negative concertation. This violates the bound on `conc_mass_comp` resulting in an infeasible solution.
2. Model verification has been done using a limited set of data from EXPOsan, however the available datasets did not include N2, NO3 or phosphate containing organisms thus those reactions have not been fully verified.
3. There is a small discrepancy between the results from EXPOsan, but this is believed to be due to a possible error in EXPOsan regarding the stoichiometric coefficients for alkalinity. Stoichiometric coefficients were verified against the original ASM2d paper.

## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
4. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/property_models/activated_sludge/asm2d_properties.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+    LiquidPhase,
+    Component,
+    Solute,
+    Solvent,
+)
+from idaes.core.util.model_statistics import degrees_of_freedom
+from idaes.core.util.initialization import fix_state_vars, revert_state_vars
+import idaes.logger as idaeslog
+
+# Some more inforation about this module
```
Comment (yalinli2) [inline watertap/property_models/activated_sludge/asm2d_properties.py]: minor typo: information (also in following modules)

File: watertap/property_models/activated_sludge/asm2d_reactions.py [hunk 2]
```diff
@@ -0,0 +1,1529 @@
...
+        # R8: Fermentation
+        # R9: Lysis
+        # R10: Storage of X_PHA
+        # R11: Aerobic storage of X_PP
+        # R12: Anoxic storage of X_PP
+        # R13: Aerobic growth of X_PAO
+        # R14: Anoxic growth of X_PAO
+        # R15: Lysis of X_PAO
+        # R16: Lysis of X_PP
+        # R17: Lysis of X_PAH
```
Comment (yalinli2) [inline watertap/property_models/activated_sludge/asm2d_reactions.py]: X_PHA?

File: watertap/property_models/activated_sludge/asm2d_reactions.py [hunk 3]
```diff
@@ -0,0 +1,1529 @@
...
+            ("R20", "Liq", "X_H"): 0,
+            ("R20", "Liq", "X_I"): 0,
+            ("R20", "Liq", "X_MeOH"): -3.45,
+            ("R20", "Liq", "X_MeP"): 4.87,
+            ("R20", "Liq", "X_PAO"): 0,
+            ("R20", "Liq", "X_PHA"): 0,
+            ("R20", "Liq", "X_PP"): 0,
+            ("R20", "Liq", "X_S"): 0,
+            ("R20", "Liq", "X_TSS"): 1.42,
+            # R21: Re-disolution
```
Comment (yalinli2) [inline watertap/property_models/activated_sludge/asm2d_reactions.py]: dissolution?

File: watertap/property_models/activated_sludge/asm2d_reactions.py [hunk 4]
```diff
@@ -0,0 +1,1529 @@
...
+                        to_units=pyo.units.kg / pyo.units.m**3 / pyo.units.s,
+                    )
+                else:
+                    return b.reaction_rate[r] == 0
+                    raise BurntToast()
+
+            self.rate_expression = pyo.Constraint(
+                self.params.rate_reaction_idx,
+                rule=rate_expression_rule,
+                doc="ASM1 rate expressions",
```
Comment (yalinli2) [inline watertap/property_models/activated_sludge/asm2d_reactions.py]: ASM2d?

File: watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py [hunk 5]
```diff
@@ -0,0 +1,623 @@
...
+        m.fs.R1.inlet.conc_mass_comp[0, "X_PP"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_PHA"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_AUT"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_MeOH"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_MeP"].fix(EPS * units.mg / units.liter)
+        # No data on TSS from EXPOsan at this point
+        m.fs.R1.inlet.conc_mass_comp[0, "X_TSS"].fix(EPS * units.mg / units.liter)
+
+        # Alkalinity was givien in mg/L based on C
+        m.fs.R1.inlet.alkalinity[0].fix(60 / 12 * units.mmol / units.liter)
```
Comment (yalinli2) [inline watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py]: 61?

File: watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py [hunk 6]
```diff
@@ -0,0 +1,623 @@
...
+        m.fs.R1.inlet.conc_mass_comp[0, "X_PHA"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_AUT"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_MeOH"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_MeP"].fix(EPS * units.mg / units.liter)
+        # No data on TSS from EXPOsan at this point
+        # However, TSS is needed for this reaction
+        m.fs.R1.inlet.conc_mass_comp[0, "X_TSS"].fix(100 * units.mg / units.liter)
+
+        # Alkalinity was givien in mg/L based on C
+        m.fs.R1.inlet.alkalinity[0].fix(60 / 12 * units.mmol / units.liter)
```
Comment (yalinli2) [inline watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py]: 61?

File: watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py [hunk 7]
```diff
@@ -0,0 +1,623 @@
...
+        model.fs.R1.inlet.conc_mass_comp[0, "S_O2"].unfix()
+        model.fs.R1.outlet.conc_mass_comp[0, "S_O2"].fix(2 * units.mg / units.liter)
+
+        solver = get_solver()
+        results = solver.solve(model, tee=True)
+        assert check_optimal_termination(results)
+
+    @pytest.mark.component
+    def test_solution(self, model):
+        # EXPOsan calculations appear to be slightly off from this implementation
```
Comment (yalinli2) [inline watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py:419]: By "off" do you mean the small discrepancies like 13.443 vs. 13.357 (for SA) - I would think it's close enough and might be caused by cumulative numeric errors (i.e., by having slightly different kinetic stoichiometry due to rounding)

**Review by yalinli2 — COMMENTED** [review]
Only minor comments, thanks @andrewlee94 !

File: watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py (hunk 7 above)
Comment (andrewlee94) [reply watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py:419]: That is what I was referring to, however it stems primarily from the comparison of the stoichiometric coefficients themselves. Before checking the predictions in an actual reactor, I did a through comparison of the stoichiometric coefficients and they all matched with the exception of the coefficients for alkalinity (which I raised with you separately). Seeing as alkalinity contributes to most of the rate expressions, any differences in the stoichiometric coefficient for alkalinity will result in differences in the rates and thus a compounding of errors.

This is almost definitely not due to rounding (unless EXPOSan is doing some significant rounding internally), as there is no rounding applies on the WaterTAP end (everything is done to machine precision), and a 1% error is far too large to be considered close.

File: watertap/property_models/activated_sludge/asm2d_reactions.py [hunk 8]
```diff
@@ -0,0 +1,1529 @@
...
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and
+# license information.
+#################################################################################
+"""
+ASM2d reaction package.
+
+Important Note: ASM2d reactions depend on the presences of TSS in solution, however
+TSS does not take part in the rate expressiosn. Thus, it is possible to have cases
+where there is insufficient TSS present in the system for the reactions ot occur
```
Comment (tarnold17) [inline watertap/property_models/activated_sludge/asm2d_reactions.py]: Nit, but change "ot" to "to"

File: watertap/property_models/activated_sludge/asm2d_reactions.py [hunk 9]
```diff
@@ -0,0 +1,1529 @@
...
+resulting in an infeasible solution,
```
Comment (tarnold17) [inline watertap/property_models/activated_sludge/asm2d_reactions.py]: End with a period, not a comma.

File: watertap/property_models/activated_sludge/asm2d_reactions.py [hunk 10]
```diff
@@ -0,0 +1,1529 @@
...
+            None
+        """
+        init_log = idaeslog.getInitLogger(blk.name, outlvl, tag="properties")
+        init_log.info("Initialization Complete.")
+
+
+@declare_process_block_class("ASM2dReactionBlock", block_class=_ASM2dReactionBlock)
+class ASM2dReactionBlockData(ReactionBlockDataBase):
+    """
+    ReactionBlcok for ASM2d.
```
Comment (tarnold17) [inline watertap/property_models/activated_sludge/asm2d_reactions.py]: Change "ReactionBlcok" to "ReactionBlock".

File: watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py [hunk 11]
```diff
@@ -0,0 +1,623 @@
...
+        # It is supected that this is due ot an error in the EXPOsan stoichiometric
```
Comment (tarnold17) [inline watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py]: "ot" -> "to". Also on line 559 below.

File: watertap/property_models/activated_sludge/asm2d_reactions.py [hunk 12]
```diff
@@ -0,0 +1,1537 @@
...
+        # R12: Anoxic storage of X_PP
+        # R13: Aerobic growth of X_PAO
+        # R14: Anoxic growth of X_PAO
+        # R15: Lysis of X_PAO
+        # R16: Lysis of X_PP
+        # R17: Lysis of X_PHA
+        # R18: Aerobic growth of X_AUT
+        # R19: Lysis of X_AUT
+        # R20: Precipitation
+        # R21: Re-disolution
```
Comment (adam-a-a) [inline watertap/property_models/activated_sludge/asm2d_reactions.py]:
```suggestion
        # R21: Re-dissolution
```

File: watertap/property_models/activated_sludge/asm2d_reactions.py [hunk 13]
```diff
@@ -0,0 +1,1537 @@
...
+        Initialization routine for reaction package.
+
+        Keyword Arguments:
+            outlvl : sets output level of initialization routine
+
+        Returns:
+            None
+        """
+        init_log = idaeslog.getInitLogger(blk.name, outlvl, tag="properties")
+        init_log.info("Initialization Complete.")
```
Comment (adam-a-a) [inline watertap/property_models/activated_sludge/asm2d_reactions.py:1020]: Is this actually initializing anything? And should this be named `initialize_build`?

File: watertap/property_models/activated_sludge/asm1_reactions.py [hunk 14]
```diff
@@ -196,8 +196,7 @@ def build(self):
         # This is the stoichiometric part the Peterson matrix in dict form
         # Note that reaction stoichiometry is on a mass basis.
         # For alkalinity, this requires converting the mass of nitrogen species
-        # reacted to be converted to mass of alkalinity converted using a
-        # charge balance (efficively MW_C/MW_N)
+        # reacted to mass of alkalinity converted using a charge balance (efficively MW_C/MW_N)
```
Comment (adam-a-a) [inline watertap/property_models/activated_sludge/asm1_reactions.py]:
```suggestion
        # reacted to mass of alkalinity converted using a charge balance (effectively MW_C/MW_N)
```

File: watertap/property_models/activated_sludge/asm2d_properties.py [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+            doc="Nitrate plus nitrite nitrogen (N03' + N02' -N). SN03 is assumed to include nitrate as well as nitrite nitrogen. [kg N/m^3]"
+        )
+        self.S_O2 = Solute(doc="Dissolved oxygen. [kg O2/m^3]")
+        self.S_PO4 = Solute(
+            doc="Inorganic soluble phosphorus, primarily ortho-phosphates. [kg P/m^3]"
+        )
+        self.S_ALK = Component(doc="Alkalinity, [mol HCO3- per m^3]")
+
+        # Particulate species
+        self.X_AUT = Solute(doc="Nitrifying organisms. [kg COD/m^3]")
```
Comment (adam-a-a) [inline watertap/property_models/activated_sludge/asm2d_properties.py]:
```suggestion
        self.X_AUT = Solute(doc="Autotrophic nitrifying organisms. [kg COD/m^3]")
```

File: watertap/property_models/activated_sludge/asm2d_properties.py [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+        self.X_H = Solute(doc="Heterotrophic organisms. [kg COD/m^3]")
+        self.X_I = Solute(doc="Inert particulate organic material. [kg COD/m^3]")
+        self.X_MeOH = Solute(doc="Metal-hydroxides. [kg TSS/m^3]")
+        self.X_MeP = Solute(doc="Metal-phosphate, MeP04. [kg TSS/m^3]")
+        self.X_PAO = Solute(doc="Phosphate-accumulating organisms. [kg COD/m^3]")
+        self.X_PHA = Solute(
+            doc="A cell internal storage product of phosphorus-accumulating organisms. [kg COD/m^3]"
```
Comment (adam-a-a) [inline watertap/property_models/activated_sludge/asm2d_properties.py]:
```suggestion
            doc="A cell internal storage product of phosphorus-accumulating organisms, primarily comprising poly-hydroxy-alkanoates (PHA). [kg COD/m^3]"
```

File: watertap/property_models/activated_sludge/asm2d_properties.py [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+            doc="Alkalinity in molar concentration",
+            units=pyo.units.kmol / pyo.units.m**3,
+        )
+
+    def get_material_flow_terms(b, p, j):
+        if j == "H2O":
+            return b.flow_vol * b.params.dens_mass
+        elif j == "S_ALK":
+            # Convert moles of alkalinity to mass assuming all is HCO3-
+            return b.flow_vol * b.alkalinity * (61 * pyo.units.kg / pyo.units.kmol)
```
Comment (adam-a-a) [inline watertap/property_models/activated_sludge/asm2d_properties.py:316]: Add a param for bicarbonate molecular weight?

**Review by yalinli2 — APPROVED** [review]
Looks good, thanks!

**Review by tarnold17 — APPROVED** [review]

**Review by adam-a-a — CHANGES_REQUESTED** [review]
Subtle changes to doc strings that I will implement. Submitting "Request Changes" to prevent merge before I implement

File: watertap/property_models/activated_sludge/asm2d_reactions.py [hunk 18]
```diff
@@ -0,0 +1,1537 @@
...
+        )
+
+
+class _ASM2dReactionBlock(ReactionBlockBase):
+    """
+    This Class contains methods which should be applied to Reaction Blocks as a
+    whole, rather than individual elements of indexed Reaction Blocks.
+    """
+
+    def initialize(blk, outlvl=idaeslog.NOTSET, **kwargs):
```
Comment (adam-a-a) [inline watertap/property_models/activated_sludge/asm2d_reactions.py:1009]:
```suggestion
    def initialize_build(blk, outlvl=idaeslog.NOTSET, **kwargs):
```

File: watertap/property_models/activated_sludge/asm2d_properties.py [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
+# by the software owners: The Regents of the University of California, through
+# Lawrence Berkeley National Laboratory,  National Technology & Engineering
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia University
+# Research Corporation, et al.  All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and
+# license information.
+#################################################################################
+"""
+Thermophysical property package to be used in conjucntion with ASM2d reactions.
```
Comment (adam-a-a) [inline watertap/property_models/activated_sludge/asm2d_properties.py]:
```suggestion
Thermophysical property package to be used in conjunction with ASM2d reactions. 

Reference:
[1] Henze, M., Gujer, W., Mino, T., Matsuo, T., Wentzel, M.C., Marais, G.v.R.,
Van Loosdrecht, M.C.M., "Activated Sludge Model No.2D, ASM2D", 1999,
Wat. Sci. Tech. Vol. 39, No. 1, pp. 165-182

```

File: watertap/property_models/activated_sludge/asm2d_reactions.py [hunk 20]
```diff
@@ -0,0 +1,1537 @@
...
+                elif r == "R20":
+                    # R20: Precipitation
+                    return b.reaction_rate[r] == pyo.units.convert(
+                        b.params.k_pre
+                        * b.conc_mass_comp_ref["S_PO4"]
+                        * b.conc_mass_comp_ref["X_MeOH"],
+                        to_units=pyo.units.kg / pyo.units.m**3 / pyo.units.s,
+                    )
+                elif r == "R21":
+                    # R21: Re-disolution
```
Comment (adam-a-a) [inline watertap/property_models/activated_sludge/asm2d_reactions.py]:
```suggestion
                    # R21: Re-dissolution
```

File: watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py [hunk 21]
```diff
@@ -0,0 +1,628 @@
...
+
+        # NOTE: Concentrations of exactly 0 result in singularities, use EPS instead
+        EPS = 1e-8
+
+        # Feed conditions based on manual mass balance of inlet and recycle streams
+        m.fs.R1.inlet.flow_vol.fix(18446 * units.m**3 / units.day)
+        m.fs.R1.inlet.temperature.fix(298.15 * units.K)
+        m.fs.R1.inlet.pressure.fix(1 * units.atm)
+        # For aerobic operation, the final spec on O2 will be on the outlet concentration
+        # This is is to account for O2 addition under aerobic operation
```
Comment (adam-a-a) [inline watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py]:
```suggestion
        # This is to account for O2 addition under aerobic operation
```

File: watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py [hunk 22]
```diff
@@ -0,0 +1,628 @@
...
+        m.fs.R1.inlet.conc_mass_comp[0, "X_PP"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_PHA"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_AUT"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_MeOH"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_MeP"].fix(EPS * units.mg / units.liter)
+        # No data on TSS from EXPOsan at this point
+        # However, TSS is needed for this reaction
+        m.fs.R1.inlet.conc_mass_comp[0, "X_TSS"].fix(100 * units.mg / units.liter)
+
+        # Alkalinity was givien in mg/L based on C
```
Comment (adam-a-a) [inline watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py]:
```suggestion
        # Alkalinity was given in mg/L based on C
```

File: watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py [hunk 23]
```diff
@@ -0,0 +1,628 @@
...
+    def test_dof(self, model):
+        assert degrees_of_freedom(model) == 0
+
+    @pytest.mark.component
+    def test_unit_consistency(self, model):
+        assert_units_consistent(model) == 0
+
+    @pytest.mark.component
+    def test_solve(self, model):
+        model.fs.R1.initialize(optarg={"bound_push": 1e-8, "mu_init": 1e-8})
```
Comment (adam-a-a) [inline watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py:553]: does model have trouble initializing/solving without these solver options? I forget what the defaults are for ipopt-watertap...not sure, if they are the same as what you have here. @bknueven do you recall what the defaults are off the top of your head? I know we can find these values plugins probably.

**Review by adam-a-a — APPROVED** [review]
LGTM. We might need to revisit this to bolster scaling as needed.

**Review by adam-a-a — APPROVED** [review]
LGTM. We might need to revisit this to bolster scaling as needed.

File: watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py (hunk 23 above)
Comment (bknueven) [reply watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py:553]: The default for `bound_push` is `1e-2` and for `mu_init` is `1e-1`.

I would prefer we avoid custom options in watertap, but in this particular case these make sense if you essentially expect to be at the solution already and just want to verify it.

**Pull request merged**